### PR TITLE
Add 6 new rules

### PR DESCRIPTION
**Pull request by jianyexi**

### CODE REVIEW

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+
+  private addToMap(map: Map<string, string[]>, key: string, value: string) {
+    if (map.has(key)) {
+      map.set(key, map.get(key).concat(value))
+    } else {
+      map.set(key, [value])
+    }
+  }
+
+  private getSpecificOperationModels(httpVerb, code) {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:60]: httpVerb means? httpMethod?

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+  }
+
+  private getResourceByName(modelName: string) {
+    if (!modelName) {
+      return undefined
+    }
+    return this.innerDoc.definitions[modelName]
+  }
+
+  private checkResource(modelName: string) {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:108]: Pls add comments here.

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+        return [...pre, cur]
+      } else {
+        return pre
+      }
+    }, [])
+    return AllOfResources
+  }
+
+  public getAllOperationGetResponseModels() {
+    return this.getSpecificOperationModels("get", "200")
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:156]: Why hard code 'get' and 200

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+
+  private getOperationGetResponseResources() {
+    const models = [...this.getAllOperationGetResponseModels().entries()].filter(m => this.checkResource(m[0]))
+    return new Map(models)
+  }
+
+  /**
+   * the result contains `put` resources
+   */
+  private getAllOperationsModels() {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:167]: getAllOperationsModelFromPut .. more clear

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+  public getAllNestedResources() {
+    const fullModels = this.getOperationGetResponseResources()
+    const nestedModels = new Set<string>()
+    for (const entry of fullModels.entries()) {
+      const paths = entry[1]
+      paths
+        /* some paths not begin with /subscriptions/, they should be ignored */
+        .filter(p => p.toLowerCase().startsWith("/subscriptions/"))
+        .some(p => {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:182]: Why use some here?

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+  }
+  /**
+   *
+   * @param path
+   * case 1 : '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/applicationGateways'
+   * return ["applicationGateways"]
+   * case 2: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/expressRouteCircuits/{circuitName}/peerings/{peeringName}'
+   * return ["expressRouteCircuits","peerings"]
+   */
+  private getResourcesTypeHierarchy(path: string) {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:224]: Do you have unit test for this function?

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+    const index = path.lastIndexOf("/providers/")
+    if (index === -1) {
+      return []
+    }
+    const lastProvider = path.substr(index)
+    const result = []
+    const matches = lastProvider.match(this.SpecificResourcePathRegEx)
+    if (matches && matches.length) {
+      let match = matches[0]
+      while (match.indexOf("/{") !== -1) {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:234]: Can filter by Regex?

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts (hunk 1 above)
Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:60]: right

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts (hunk 7 above)
Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:234]: synced offline

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts (hunk 6 above)
Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:224]: private function

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts (hunk 5 above)
Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:182]: synced offline

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts (hunk 4 above)
Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:167]: synced offline

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts (hunk 2 above)
Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:108]: added

File: src/typescript/azure-openapi-validator/rules/AllResourcesMustHaveGetOperation.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    */
+    const utils = new ResourceUtils(doc)
+    const allCollectionModels = utils.getCollectionModels()
+    const allCollectionApiInfo = utils.getCollectionApiInfo()
+    const allResourcesHavingGetOperation = utils.getAllOperationGetResponseModels()
+    for (const resource of allCollectionModels.entries()) {
+      if (
+        allResourcesHavingGetOperation.has(resource[0]) &&
+        !allResourcesHavingGetOperation.has(resource[1]) &&
+        !allCollectionApiInfo.some(info => info.modelName === resource[0])
```
Comment (jianyexi) [inline src/typescript/azure-openapi-validator/rules/AllResourcesMustHaveGetOperation.ts:30]: add some comments

File: src/typescript/azure-openapi-validator/rules/OperationsApiResponseSchema.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  category: "ARMViolation",
+  mergeState: MergeStates.composed,
+  openapiType: OpenApiTypes.arm,
+
+  *run(doc, node, path) {
+    const msg: string = 'The response schema of operations API "{0}" does not match the ARM specification. Please standardize the schema.'
+    /**
+     * 1 get the operations API and schema
+     * 2 verify the schema
+     */
```
Comment (jianyexi) [inline src/typescript/azure-openapi-validator/rules/OperationsApiResponseSchema.ts:22]: add ARM document

**Pull request merged**